libc/src/__support/fixed_point/fx_bits.h

```diff
@@ -142,6 +160,25 @@ template <typename T> LIBC_INLINE constexpr T abs(T x) {
   }
 }
 
+template <typename T>
+LIBC_INLINE constexpr cpp::enable_if_t<cpp::is_fixed_point_v<T>, int>
+countls(T x) {
+  using FXRep = FXRep<T>;
+  using BitType = typename FXRep::StorageType;
+  constexpr int CONTAIN_LEN = cpp::numeric_limits<BitType>::digits;
+  constexpr int PADDING_LEN = CONTAIN_LEN - FXRep::VALUE_LEN;
+
+  if constexpr (FXRep::SIGN_LEN != 0) {
+    if (x < 0) {
+      x = -(x + FXRep::EPS());
```

Comment:
isn't this just take the complement of the bits? Does the operator `~` work for fixed point types?

Reply:
> 4.1.6.1.2
> ...
> It is not allowed to apply the complement
operator (~) to a fixed-point operand. 
> ...

Reply:
That's an odd one, something maybe we should revise when we try to add fixed point types to C standard.  I guess they were worried about integer promotion rules.  Anyway, you can try:
```
  x = FXBits(static_cast<StorageType>(~value)).get_val();
```
to see if it works properly.

Reply:
After thinking about it. You're right. This is equivalent to bitwise complement. 

I also found this function in `fx_bits.h`.

https://github.com/llvm/llvm-project/blob/f71e13eb354d6a02710ceec254e8c399c60ffdbf/libc/src/__support/fixed_point/fx_bits.h#L143-L150

Assuming that `static_cast` and `bit_cast` are free, this is one less operation than the current way I do it.

So `static_cast` should be free, as it seems its only used as a type hint here.
I think `bit_cast` is free by definition.

Reply:
This is fixed in [cc87a9a](https://github.com/llvm/llvm-project/commit/cc87a9ac973131db1047a22ebfb2b0db2d2f1926).

https://github.com/llvm/llvm-project/blob/cc87a9ac973131db1047a22ebfb2b0db2d2f1926/libc/src/__support/fixed_point/fx_bits.h#L165-L168